File: 0912-sort-an-array/0912-sort-an-array.java
class Solution {
    public int[] sortArray(int[] nums) {
        if(nums == null || nums.length <= 1){
            return nums;
        }
        mergeSort(nums, 0, nums.length - 1, new int[nums.length]);
        return nums;
    }

    private void mergeSort(int[] a, int l, int r, int[] temp){
        if(l >= r) return;
        int mid = l + (r - l) / 2;
        mergeSort(a, l, mid, temp);
        mergeSort(a, mid + 1, r, temp);
        merge(a, l, mid, r, temp);
    }

    private void merge(int[] a, int l, int mid, int r, int[] temp){
        int i = l, j = mid + 1, k = l;
        while(i <= mid && j <= r){
            if(a[i] <= a[j]){
                temp[k++] = a[i++];
            }else{
                temp[k++] = a[j++];
            }
        }
        while(i <= mid) temp[k++] = a[i++];
        while(j <= r) temp[k++] = a[j++];
        for(k = l; k <= r; k++){
            a[k] = temp[k];
        }
    }
}

File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:

        n = len(nums)

        def heapify(i, size):
            while True:
                largest = i
                left = 2 * i + 1
                right = 2 * i + 2

                if left < size and nums[left] > nums[largest]:
                    largest = left

                if right < size and nums[right] > nums[largest]:
                    largest = right

                if largest == i:
                    break

                nums[i], nums[largest] = nums[largest], nums[i]
                i = largest

        # Build max heap
        for i in range(n // 2 - 1, -1, -1):
            heapify(i, n)

        # Move largest element to the end
        for end in range(n - 1, 0, -1):
            nums[0], nums[end] = nums[end], nums[0]
            heapify(0, end)

        return nums